### PR TITLE
[CARBONDATA-1281] Support multiple temp dirs for writing files while loading data

### PR DESCRIPTION
**Pull request by xuchuanyin**
# Modifications
This feature mainly focus on avoiding disk hot-spot in single massive data loading, changes are made in two parts: 

1. randomly choose a yarn local folder while writing sort temp file each time in sort-process;

2.randomly choose a yarn local folder while writing carbondata file each time in write-process.

# Usage

To enable this feature, user should enable `carbon.use.multi.temp.dir=true` and `carbon.use.local.dir=true`.

# Performance
In my case, this feature improves the loading performance from 35M/s/node to 70+M/s/node

### CODE REVIEW

**Comment by asfgit** [comment]
Can one of the admins verify this patch?

**Comment by asfgit** [comment]
Can one of the admins verify this patch?

**Comment by xuchuanyin** [comment]
I created this PR and closed #1195

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 1.6, Please check CI http://144.76.159.231:8080/job/ApacheCarbonPRBuilder/604/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/3199/

**Comment by asfgit** [comment]
Can one of the admins verify this patch?

**Comment by asfgit** [comment]
Can one of the admins verify this patch?

**Comment by xuchuanyin** [comment]
There is no useful information in the compilation message.

retest this please

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/3200/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 1.6, Please check CI http://144.76.159.231:8080/job/ApacheCarbonPRBuilder/605/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/3201/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 1.6, Please check CI http://144.76.159.231:8080/job/ApacheCarbonPRBuilder/606/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6, Please check CI http://144.76.159.231:8080/job/ApacheCarbonPRBuilder/608/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/3203/

**Comment by chenliang613** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/3205/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 1.6, Please check CI http://144.76.159.231:8080/job/ApacheCarbonPRBuilder/610/

**Comment by chenliang613** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6, Please check CI http://144.76.159.231:8080/job/ApacheCarbonPRBuilder/611/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/3206/

File: core/src/main/java/org/apache/carbondata/core/constants/CarbonCommonConstants.java [hunk 1]
```diff
@@ -1296,6 +1296,18 @@
...
+   * whether to use multi directories when loading data,
+   * the main purpose is to avoid single-disk-hot-spot
+   */
+  @CarbonProperty
+  public static final String CARBON_USE_MULTI_TEMP_DIR = "carbon.use.multiple.temp.dir";
+
+  /**
+   * default value for multi temp dir
+   */
+  public static final String CARBON_USING_MULTI_TEMP_DIR_DEFAULT = "false";
```
Comment (xuchuanyin) [inline core/src/main/java/org/apache/carbondata/core/constants/CarbonCommonConstants.java]: :+1: fixed

File: processing/src/main/java/org/apache/carbondata/processing/util/CarbonDataProcessorUtil.java [hunk 2]
```diff
@@ -145,21 +146,31 @@ public static void renameBadRecordsFromInProgressToNormal(
...
     }
   }
 
   /**
+   * This method will be used to create dirs
+   * @param locations locations to create
+   */
+  public static void createLocations(String[] locations) {
+    for (String loc : locations) {
+      if (new File(loc).mkdirs()) {
```
Comment (xuchuanyin) [inline processing/src/main/java/org/apache/carbondata/processing/util/CarbonDataProcessorUtil.java]: :+1: nice

**Comment by xuchuanyin** [comment]
All review comments solved

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6, Please check CI http://144.76.159.231:8080/job/ApacheCarbonPRBuilder/615/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/3210/

**Comment by xuchuanyin** [comment]
@chenliang613 Does this PR can be merged or need more reviews?

**Comment by chenliang613** [comment]
LGTM, very good PR! Thanks for your good contribution.

**Comment by xuchuanyin** [comment]
There is a bugfix for this PR #1206

**Pull request closed without merging**